eureka-core/src/main/java/com/netflix/eureka/registry/AbstractInstanceRegistry.java

```diff
@@ -1278,29 +1280,113 @@ long getCurrentTimeNano() {  // for testing
 
     }
 
-    private class CircularQueue<E> extends ConcurrentLinkedQueue<E> {
+    /* visible for testing */  static class CircularQueue<E> extends AbstractQueue<E> {
```

Comment:
This looks correct to me but considering you're rolling out a data structure, the amount of unit tests is expected to be much larger, I don't see test cases for [0, 1, even #, odd #] capacity buffers, tests for ops when under capacity, same tests for at capacity, etc. The risk/benefit ratio is just too high.

Based on your benchmarks and iterations so far, it seems like the best way forward is to
- Changed to LinkedList-based as in [iteration 1](https://github.com/Netflix/eureka/pull/1237/commits/e08ddfa7ba2f3b1e296707c033629dbd4b760f24)
- Fix `clearRegistry` to be thread safe as you did [here](https://github.com/Netflix/eureka/pull/1237/commits/3e7a8246827a1abea80e55c9da1a7f4aa3eaa479#diff-f80bacaf7859f75ae25a290673be6b19R174)
- Pre-allocate result lists in `getLastN ` methods similar to https://github.com/Netflix/eureka/pull/1237/commits/3e7a8246827a1abea80e55c9da1a7f4aa3eaa479#diff-f80bacaf7859f75ae25a290673be6b19R174

This would be correct, about the same performance, get rid of unnecessary synchronization, require no additional testing.

Let me know what you think.